autogen/agentchat/groupchat.py

```diff
@@ -496,6 +528,299 @@ def _finalize_speaker(self, last_speaker: Agent, final: bool, name: str, agents:
         agent = self.agent_by_name(name)
         return agent if agent else self.next_agent(last_speaker, agents)
 
+    def _auto_select_speaker(
+        self,
+        last_speaker: Agent,
+        selector: ConversableAgent,
+        messages: Optional[List[Dict]],
+        agents: Optional[List[Agent]],
+    ) -> Agent:
+        """Selects next speaker for the "auto" speaker selection method. Utilises its own two-agent chat to determine the next speaker and supports requerying.
+
+        Speaker selection for "auto" speaker selection method:
+        1. Create a two-agent chat with a speaker selector agent and a speaker validator agent, like a nested chat
+        2. Inject the group messages into the new chat
+        3. Run the two-agent chat, evaluating the result of response from the speaker selector agent:
+            - If a single agent is provided then we return it and finish. If not, we add an additional message to this nested chat in an attempt to guide the LLM to a single agent response
+        4. Chat continues until a single agent is nominated or there are no more attempts left
+        5. If we run out of turns and no single agent can be determined, the next speaker in the list of agents is returned
+
+        Args:
+            last_speaker Agent: The previous speaker in the group chat
+            selector ConversableAgent:
+            messages Optional[List[Dict]]: Current chat messages
+            agents Optional[List[Agent]]: Valid list of agents for speaker selection
+
+        Returns:
+            Dict: a counter for mentioned agents.
+        """
+
+        # If no agents are passed in, assign all the group chat's agents
+        if agents is None:
+            agents = self.agents
+
+        # The maximum number of speaker selection attempts (including requeries)
+        # We track these and use them in the validation function as we can't
+        # access the max_turns from within validate_speaker_name
+        max_attempts = 1 + self.max_retries_for_selecting_speaker
```

Comment:
Maybe a short comment on why we need to `1 + self.max_retries_for_selecting_speaker`?